File: php/63-unique-paths2.php
<?php
class Solution {

    /**
     * @param Integer[][] $obstacleGrid
     * @return Integer
     */
    function uniquePathsWithObstacles($obstacleGrid) {
        $m = count($obstacleGrid);
        $n = count($obstacleGrid[0]);
        
        $counts[0][0] = 1;
        for ($i = 0; $i < $m; $i++) {
            for ($j = 0; $j < $n; $j++) {
                if ($obstacleGrid[$i][$j] == 1) {
                    $counts[$i][$j] = 0;
                    continue;
                }
                $left = $counts[$i-1][$j]?? 0;
                $up = $counts[$i][$j-1]?? 0;
                if (!isset($counts[$i][$j])) {
                    $counts[$i][$j] = $left + $up;
                }
            }
        }
        return $counts[$m-1][$n-1];
    }
}

File: go/63-unique-paths2.go
package main

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	m := len(obstacleGrid)
	n := len(obstacleGrid[0])

	record := make(map[int]map[int]int, m)
	for i := 0; i < m; i++ {
		record[i] = make(map[int]int, n)
	}
	record[0][0] = 1
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			left := 0
			up := 0
			if obstacleGrid[i][j] == 1 {
				record[i][j] = 0
				continue
			}
			if l, ok := record[i-1][j]; ok {
				left = l
			}
			if u, ok := record[i][j-1]; ok {
				up = u
			}
			if left == 0 && up == 0 {
				continue
			}
			record[i][j] = left + up
		}
	}
	return record[m-1][n-1]
}
